### PR TITLE
Airdrop Contract

### PR DESCRIPTION
**Pull request by arhamchordia**
## 1. Overview

This PR aims to introduce Airdrop contract on Quasar chain which will serve as the contract for releasing airdrops.

## 2. Implementation details

- It contains the state file for the contract (from the spec doc)
- Instantiate message that takes the airdrop config (from spec doc)
- Admin messages: 
    * Update config (acts as modifier and starter of the airdrop)
    * Add users
    * Set users
    * Remove users
    * Withdraw funds
- User side message: 
    * Claim 
- Queries:
    * Query config
    * Query a user info
    * Query contract state
    * Query airdrop readiness
    * Query airdrop specs

## 3. How to test/use

- There is shell script test that can be run in order to test the state changes in the contract. 

## 4. Checklist

- [ ] Does the Readme need to be updated?

### CODE REVIEW

File: smart-contracts/contracts/airdrop/src/admin.rs
Comment (njerschow) [inline smart-contracts/contracts/airdrop/src/admin.rs:79]: never true because Uint

Comment (njerschow) [reply smart-contracts/contracts/airdrop/src/admin.rs:79]: can still verify == zero() i think

File: smart-contracts/contracts/airdrop/src/users.rs
Comment (njerschow) [inline smart-contracts/contracts/airdrop/src/users.rs:35]: same issue with loading contract balance

File: smart-contracts/contracts/airdrop/src/users.rs
Comment (njerschow) [inline smart-contracts/contracts/airdrop/src/users.rs:39]: redundant but still good.

File: smart-contracts/contracts/airdrop/src/users.rs
Comment (njerschow) [inline smart-contracts/contracts/airdrop/src/users.rs:45]: we don't do anything with the transfer message. we need to pass it as add_message to response for it to execute

File: smart-contracts/contracts/airdrop/src/admin.rs
Comment (LaurensKubat) [inline smart-contracts/contracts/airdrop/src/admin.rs:23]: These block height checks are hard to read, a comment on what you're trying to achieve with them would be, such as 
```
start_height and end_height should always be in the future
```

File: smart-contracts/contracts/airdrop/src/admin.rs
Comment (LaurensKubat) [inline smart-contracts/contracts/airdrop/src/admin.rs:20]: Is this check needed together with the check on line 23?

File: smart-contracts/contracts/airdrop/src/admin.rs
Comment (LaurensKubat) [inline smart-contracts/contracts/airdrop/src/admin.rs:115]: Does this require us to set a new airdrop config set in the future and then seperately update the user set?

File: smart-contracts/contracts/airdrop/src/admin.rs
Comment (LaurensKubat) [inline smart-contracts/contracts/airdrop/src/admin.rs:201]: Todo should be added here, especially since `user_info.get_claimed_flag() == true` leads to no removal but an Ok response

**Review by LaurensKubat — COMMENTED** [review]
General idea looks good. My main question is, why do we have a single set of users to which all airdrops are applicable. I expect this will introduce problems where we try to airdrop multiple airdrops to different groups of users. My suggestion would be to  either have 1 contract for 1 airdrop with 1 set of users, or have multiple airdrops in the contract but make each airdrop have their own Map of users. The first approach where we just instantiate a new contract per user would be easiest imo.

File: smart-contracts/contracts/airdrop/src/admin.rs
Comment (arhamchordia) [reply smart-contracts/contracts/airdrop/src/admin.rs]: Changed it to contract address.

File: smart-contracts/contracts/airdrop/src/admin.rs
Comment (arhamchordia) [reply smart-contracts/contracts/airdrop/src/admin.rs]: Resolved.

File: smart-contracts/contracts/airdrop/src/admin.rs
Comment (arhamchordia) [reply smart-contracts/contracts/airdrop/src/admin.rs:117]: Yes. It's a check on add and set users so that we never exceed the airdrop size.

File: smart-contracts/contracts/airdrop/src/admin.rs
Comment (arhamchordia) [reply smart-contracts/contracts/airdrop/src/admin.rs]: Resolved.

File: smart-contracts/contracts/airdrop/src/admin.rs
Comment (arhamchordia) [reply smart-contracts/contracts/airdrop/src/admin.rs:182]: Sure. This removes the redundancy.

File: smart-contracts/contracts/airdrop/src/admin.rs
Comment (arhamchordia) [reply smart-contracts/contracts/airdrop/src/admin.rs]: Resolved.

File: smart-contracts/contracts/airdrop/src/admin.rs
Comment (arhamchordia) [reply smart-contracts/contracts/airdrop/src/admin.rs:225]: The transaction is set to be executed only by Admin.

File: smart-contracts/contracts/airdrop/src/admin.rs
Comment (arhamchordia) [reply smart-contracts/contracts/airdrop/src/admin.rs]: Yes. I realised later that these messages needs to be added in order to execute them.

File: smart-contracts/contracts/airdrop/src/contract.rs
Comment (arhamchordia) [reply smart-contracts/contracts/airdrop/src/contract.rs:32]: Instantiate contains the whole config in a single var. So, that's the reason why I added this check.

File: smart-contracts/contracts/airdrop/src/users.rs
Comment (arhamchordia) [reply smart-contracts/contracts/airdrop/src/users.rs:11]: Yes. A good idea to have a separate validation.

File: smart-contracts/contracts/airdrop/src/users.rs
Comment (arhamchordia) [reply smart-contracts/contracts/airdrop/src/users.rs]: Resolved.

File: smart-contracts/contracts/airdrop/src/users.rs
Comment (arhamchordia) [reply smart-contracts/contracts/airdrop/src/users.rs]: Resolved.

File: smart-contracts/contracts/airdrop/src/admin.rs
Comment (arhamchordia) [reply smart-contracts/contracts/airdrop/src/admin.rs:22]: Sure. Adding more comments soon.

File: smart-contracts/contracts/airdrop/src/admin.rs
Comment (arhamchordia) [reply smart-contracts/contracts/airdrop/src/admin.rs]: Changed the checks a little as I wrote them in a hurry.

File: smart-contracts/contracts/airdrop/src/admin.rs
Comment (arhamchordia) [reply smart-contracts/contracts/airdrop/src/admin.rs:131]: New airdrop would mean a new contract.

File: smart-contracts/contracts/airdrop/src/admin.rs
Comment (arhamchordia) [reply smart-contracts/contracts/airdrop/src/admin.rs:221]: Oh okay. Added a separate array of only the removed users.

**Comment by arhamchordia** [comment]
> General idea looks good. My main question is, why do we have a single set of users to which all airdrops are applicable. I expect this will introduce problems where we try to airdrop multiple airdrops to different groups of users. My suggestion would be to either have 1 contract for 1 airdrop with 1 set of users, or have multiple airdrops in the contract but make each airdrop have their own Map of users. The first approach where we just instantiate a new contract per user would be easiest imo.

1 airdrop 1 contract is the idea we stuck to while writing the spec for the contract.

**Comment by akure** [comment]
General comments - 
Utility functions could be used for easy code readability, and use them wherever needed.
is_airdrop_active
is_window_initialised
is_contract_has_enough_balance

File: smart-contracts/contracts/airdrop/src/admin.rs
Comment (akure) [inline smart-contracts/contracts/airdrop/src/admin.rs:109]: Add the checks if airdrop is still active.

File: smart-contracts/contracts/airdrop/src/admin.rs
Comment (akure) [inline smart-contracts/contracts/airdrop/src/admin.rs:337]: could be divided in to two checks,
1. is_window_initialised(..) ->  ( current_airdrop_config.end_height == || current_airdrop_config.start_height == 0 )
2. is_airdrop_ended(..) ->  ( env.block.height 
3.  > current_airdrop_config.end_height )

File: smart-contracts/contracts/airdrop/src/contract.rs
Comment (akure) [inline smart-contracts/contracts/airdrop/src/contract.rs:53]: && shound'nt this be || if you want to initialise the contract with zero values of start height and end height. And msg.config.total_claimed is zero.

File: smart-contracts/contracts/airdrop/src/admin.rs
Comment (arhamchordia) [reply smart-contracts/contracts/airdrop/src/admin.rs:115]: This can only be done if both end and start height are non zero. Which means that the airdrop has been activated once. So no need to checks if the airdrop is still active.

File: smart-contracts/contracts/airdrop/src/contract.rs
Comment (LaurensKubat) [inline smart-contracts/contracts/airdrop/src/contract.rs:123]: You can shorthand this to ```AdminExecuteMsg::UpdateAirdropConfig(airdrop_config) => execute_update_airdrop_config(deps, env, airdrop_config),```

File: smart-contracts/contracts/airdrop/src/contract.rs
Comment (LaurensKubat) [inline smart-contracts/contracts/airdrop/src/contract.rs:166]: This isn't clean description of what replies do, a better way to think about replies would be callbacks after with results of msgs dispatched and executed to the chain

File: smart-contracts/contracts/airdrop/src/helpers.rs
Comment (LaurensKubat) [inline smart-contracts/contracts/airdrop/src/helpers.rs:90]: This should be a variant of `AirdropErrors`

File: smart-contracts/contracts/airdrop/src/query.rs
Comment (LaurensKubat) [inline smart-contracts/contracts/airdrop/src/query.rs:76]: Could use a different name, such as airdrop_validity or something

File: smart-contracts/contracts/airdrop/src/users.rs
Comment (LaurensKubat) [inline smart-contracts/contracts/airdrop/src/users.rs:49]: We probably don't have to use a reply here. An option to not use replies would be to add the claim transfer message as a regular message, if that fails, the entire contract call rolls back. Thus we can update the users claim to claimed before the response is sent

**Review by LaurensKubat — APPROVED** [review]
Some general comments:
- Comments such as ` // Define a helper function to create a mock contract configuration` don't necessarily add information for later readers, but might become outdated/ require later maintainers to maintain it. Removing these comments would improve the overall health of our codebase
- I added a ton of nitpicky comments. These do not have to be improved before merging, but would improve your overall code quality
- We should follow rustdoc standards for doc comments
-This contract could be written without usage of replies

File: smart-contracts/contracts/airdrop/src/contract.rs
Comment (arhamchordia) [reply smart-contracts/contracts/airdrop/src/contract.rs:123]: `cargo fmt` gives an error if I try to do it this way.

File: smart-contracts/contracts/airdrop/src/contract.rs
Comment (arhamchordia) [reply smart-contracts/contracts/airdrop/src/contract.rs:166]: Sure.

File: smart-contracts/contracts/airdrop/src/helpers.rs
Comment (arhamchordia) [reply smart-contracts/contracts/airdrop/src/helpers.rs:90]: Sure.

File: smart-contracts/contracts/airdrop/src/users.rs
Comment (arhamchordia) [reply smart-contracts/contracts/airdrop/src/users.rs:49]: Makes sense. I can get rid it right away.

File: smart-contracts/contracts/airdrop/src/admin.rs
Comment (arhamchordia) [reply smart-contracts/contracts/airdrop/src/admin.rs:131]: @njerschow suggested to do it this way. 
Plus I needed to have index in order to make debugging easier based on the error log.

File: smart-contracts/contracts/airdrop/src/admin.rs
Comment (LaurensKubat) [reply smart-contracts/contracts/airdrop/src/admin.rs:293]: @arhamchordia For future reference, please integrate this, makes reading code easier

File: smart-contracts/contracts/airdrop/src/admin.rs
Comment (LaurensKubat) [inline smart-contracts/contracts/airdrop/src/admin.rs:186]: Generally in rust, in the impl for a struct we add a new() method and call that

**Review by LaurensKubat — APPROVED** [review]
lgtm

File: smart-contracts/contracts/airdrop/src/admin.rs
Comment (arhamchordia) [reply smart-contracts/contracts/airdrop/src/admin.rs:293]: Sounds good.

**Pull request merged**
